karabo_data/geometry2.py

```diff
@@ -335,6 +335,68 @@ def to_distortion_array(self):
 
         return distortion
 
+    def _tile_slice(self, tileno):
+        """Implement in subclass: which part of module array each tile is.
+        """
+        raise NotImplementedError
+
+    def get_pixel_positions(self, centre=True):
+        """Get the coordinates of each pixel in the detector
+
+        The output is an array with shape like the data, with an extra dimension
+        of length 3 to hold (x, y, z) coordinates. Coordinates are in pixel
+        units from the detector centre.
+
+        If centre=True, the coordinates are calculated for the centre of each
+        pixel. If not, the coordinates are for the first corner of the pixel
+        (the one nearest the [0, 0] corner of the tile in data space.
+        """
+        out = np.zeros(self.expected_data_shape + (3,), dtype=np.float32)
+
+        # Prepare some arrays to use inside the loop
+        pixel_ss_index, pixel_fs_index = np.meshgrid(
+            np.arange(0, self.frag_ss_pixels),
+            np.arange(0, self.frag_fs_pixels),
+            indexing='ij'
+        )
+        if centre:
+            pixel_ss_index += 0.5
+            pixel_fs_index += 0.5
```

Comment:
I also don't understand this, pixel_ss/fs_indexes are the pixels/points in ss and fs direction on a 2D grid? Then why adding half a pixel in each direction? I mean correct me if I am wrong but shouldn't the pixels represent the smallest addressable elements on a lattice? And hence are the center by default ... well unless there is a fixed axis associated with them. I'll think about it

Reply:
If you imagine a Cartesian grid, the first pixel stretches from x=0 to x=1, and y=0 to y=1. Its indices are (0, 0), but the coordinates of its centre is (0.5,  0.5).

Calling these variables `index` is wrong, though - that was lazy copy-pasting.